build: add devcontainer configuration

bnb opened the pull request:
this PR adds two metadata files: 

1. `./.devcontainer/.devcontainer.json` provides metadata to GitHub Codespaces and other tools (like Docker Desktop's preview [Developer Environments](https://docs.docker.com/desktop/dev-environments/) feature) about how to build and configure a developer environment.
2. `./.devcontainer/Dockerfile` a very simple `Dockerfile` that the `.devcontainer.json` file uses to build from. This will consume the Docker image I've built out (pending https://github.com/nodejs/admin/issues/641 on moving it into nodejs/devcontainer) that will be built and published nightly.

These should add a relatively low maintenance burden - the `Dockerfile` is literally two LOC (most of the work for this is done in the nodejs/devcontainer repo!) and the `.devcontainer.json` is effectively a configuration file that represents some settings. It could be stripped down or more thoroughly built out if folks want, though I'd generally like to request that we err on configuring for a broader audience than just project members (think the audience at a Code and Learn event) while also serving ourselves.

This PR should not land until https://github.com/nodejs/admin/issues/641 lands.

bnb commented:
I should also add:

The intent of adding these files is to allow for easier direct development of Node.js. It means that when you try to launch Codespaces from nodejs/node in GitHub, you'd be set up with the environment defined in these configuration files rather than a raw default environment with a git clone of the source and nothing else. It also means that you'd be able to effectively do the same with Docker Desktop's Development Environment preview and any future implementors of the devcontainer config format (GitPod, for example, has this as a roadmap item).

These configurations **are not** required to use the Docker image from https://github.com/nodejs/admin/issues/641 as a developer environment. Rather, they take the image built from that repo and extend it slightly to improve the experience outside of those who want to figure out how to launch the Docker container that repo builds locally or how to spin it up as a chonky VM in the cloud and SSH in to it.

bnb commented on .devcontainer/.devcontainer.json:
```diff
@@ -0,0 +1,20 @@
+          }
+        }
+      }
+  }
```
fixed ty

bnb replied:
actually fixed now :P

F3n67u commented:
is this pr still relevant? 

I am using GitHub codespace for node source code development, but we have no devcontainer configuration to config the docker env of GitHub codespace so I have to install ninja manually.

bnb commented:
This PR is still relevant (imo). It was stalled for a long time waiting on us to get a Docker org. We now have that. Unfortunately around that time I was feeling pretty drained and was beginning to feel the need to job search. I've now job searched, quit my job, and am on a brief mental health recovery period that was graciously agreed to by my new employer before working again.

This is maybe like 5 hours of work from being complete, not including figuring out who in the org owns the Node.js Docker Hub org and setting that governance up.

Codespaces *does* work now, but kinda sucks as you mention. The Docker image has been transferred into the org but isn't being built/deployed presently. I'll be able to spend some time on this in the back half of June! Also happy to pair on it at Summit if anyone would like that.

bnb commented:
some small updates from airplane driven development:

- re-tested the Docker image to make sure it still works (it does!)
- updated the nodejs/devcontainer repo with some more documentation from my exploration of re-testing it with fresh eyes/a fresh computer
- updated the tag in nodejs/devcontainer to be `nightly` not `latest`, since it's more accurate
- updated this PR to address @targos's feedback
- tweaked the Docker image tag from `latest` to `nightly` since that's both what's in nodejs/devcontainer and a closer match to what we're actually initially doing with that repo.

bnb commented:
tl;dr update on this: 

- Code in the repo seems to work
- This addition is minimal but relies on that code
- The remaining hurdle is figuring out how to set up auth and publishing to our org (nodejs) on DockerHub, since that's a new property for us.

I'd love to try to get that last part done while in-person at OpenJS World in Austin. It should be extremely do-able. If we can unblock that, we should be good to get the nodejs/devcontainer repo automated, let it simmer for a bit and see if there are any problems, and then from there merge this PR.

bnb commented:
Tagging with TSC agenda per a discussion I had with a TSC member:

Currently, the only thing blocking this from landing (outside of the failures that shouldn't be failing?) is working with someone to get the Docker auth setup. The OpenJS Foundation staff have the keys to the org that we need, so the remaining task is just doing that manual work to get things set up. 

I'm happy to pair on this with anyone almost anytime in the next ~month since I'll have plenty of time. Just need to figure out who/when :)

mcollina commented:
With Brian leaving OpenJS Foundation I do not know who to ask, but we can ask Robin to direct us.

mcollina commented:
@nodejs/build are there any volunteers that would like to pair with Tierney?

ErickWendel commented:
> @nodejs/build are there any volunteers that would like to pair with Tierney?

✋🏻

mhdawson commented:
Removing from the TSC agenda, Tierney walked me through the setup.

bnb commented:
small status update: the repo (nodejs/devcontainer) seems to be working. It published the first container successfully. I want to let it publish nightly for at least a week without problems before considering requesting this PR be merged, but we shouldn't be far off of that now 👍🏻

bnb commented:
Just a small update: The Docker image seems to have successfully built + published for the past 23 days! 🎉 

I just followed the instructions and it also seems to be working exactly as one would hope (everyone who's going to use this might want to start working on a custom environment setup script to make it pretty how you like it):

![image](https://user-images.githubusercontent.com/502396/181589455-48b6b51a-2f64-4cce-b571-5eb82cbadebb.png)

On this PR: I need to figure out what about the setup is currently killing all actions. Once I've done that, we should be good to go.

richardlau commented:
> 
> On this PR: I need to figure out what about the setup is currently killing all actions. Once I've done that, we should be good to go.

https://github.com/nodejs/node/pull/40825/commits/4d90714416d139ce4b28625ece67f0d38389bcc3 added a Git submodule (`devcontainer`) and the action workflows are complaining about that.

bnb commented:
> https://github.com/nodejs/node/commit/4d90714416d139ce4b28625ece67f0d38389bcc3 added a Git submodule (devcontainer) and the action workflows are complaining about that.

Gotcha, I wouldn't have figured that out fast so thank you. I've literally never touched submodules so must have somehow done this accidentally and will figure out how to undo it 😅

bnb commented:
Force pushed the same changes _hopefully_ sans submodule 🤞🏻

bnb commented:
If tests pass, this should be mergable. Since this theoretically doesn't change any of Node.js itself, tests _should_ pass 👀

Trott commented:
Putting this in draft mode and then putting it back in ready-for-review so that the tests can re-run. Current tests show failures that are almost certainly unrelated.

Trott commented:
Landed in 0c5f25333961
